examples/getting_started/tools_and_agents.ipynb

```diff
@@ -0,0 +1,519 @@
+   "metadata": {},
+   "source": "For more information on setting up Mirascope and its dependencies, see the [Mirascope installation guide](https://docs.mirascope.io/latest/get-started/)."
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Understanding Tools in Mirascope\n",
+    "\n",
+    "In Mirascope, Tools are functions that extend the capabilities of LLMs. They allow the LLM to perform specific tasks or access external data. Tools are typically used to:\n",
+    "\n",
+    "1. Retrieve information from external sources\n",
+    "2. Perform calculations or data processing\n",
+    "3. Interact with APIs or databases\n",
+    "4. Execute specific actions based on the LLM's decisions\n",
+    "\n",
+    "Let's start by creating a simple Tool that extracts content from a webpage:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-09-10T05:49:48.725564Z",
+     "start_time": "2024-09-10T05:49:48.658370Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "\n",
+    "import requests\n",
+    "from bs4 import BeautifulSoup\n",
+    "\n",
+    "\n",
+    "def extract_content(url: str) -> str:\n",
+    "    \"\"\"Extract the main content from a webpage.\n",
+    "\n",
+    "    Args:\n",
+    "        url: The URL of the webpage to extract the content from.\n",
+    "\n",
+    "    Returns:\n",
+    "        The extracted content as a string.\n",
+    "    \"\"\"\n",
+    "    try:\n",
+    "        response = requests.get(url, timeout=5)\n",
+    "\n",
+    "        soup = BeautifulSoup(response.content, \"html.parser\")\n",
+    "\n",
+    "        unwanted_tags = [\"script\", \"style\", \"nav\", \"header\", \"footer\", \"aside\"]\n",
+    "        for tag in unwanted_tags:\n",
+    "            for element in soup.find_all(tag):\n",
+    "                element.decompose()\n",
+    "\n",
+    "        main_content = (\n",
+    "            soup.find(\"main\")\n",
+    "            or soup.find(\"article\")\n",
+    "            or soup.find(\"div\", class_=re.compile(\"content|main\"))\n",
+    "        )\n",
+    "\n",
+    "        if main_content:\n",
+    "            text = main_content.get_text(separator=\"\\n\", strip=True)\n",
+    "        else:\n",
+    "            text = soup.get_text(separator=\"\\n\", strip=True)\n",
+    "\n",
+    "        lines = (line.strip() for line in text.splitlines())\n",
+    "        return \"\\n\".join(line for line in lines if line)\n",
+    "    except Exception as e:\n",
+    "        return f\"{type(e)}: Failed to extract content from URL {url}\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This `extract_content` function is a Tool that takes a URL as input and returns the main content of the webpage as a string. It uses BeautifulSoup to parse the HTML and extract the relevant text content.\n",
```

Comment:
I think we need to provide more explanation here around what exactly is happening. For example, we should mention that when passed to a call it will be automatically converted into a `BaseTool` class, and when passed to the LLM API we use this schema to auto-generate the actual schema the API expects. We should likely show an example of what this schema looks like (use the `tool_schema` method and something like `OpenAITool.type_from_fn(...)` to get the correct type to output the correct schema rather than writing it by hand)